contracts/axelar-gateway/tests/test.rs

```diff
@@ -490,3 +492,346 @@ fn upgrade_unauthorized() {
+
+    // Modify signatures to make them invalid
+    let mut new_signers = Vec::new(&env);
+    for signer in proof.signers.iter() {
+        new_signers.push_back(ProofSigner {
+            signer: signer.signer,
+            signature: ProofSignature::Unsigned,
+        });
+    }
+    proof.signers = new_signers;
+
+    let new_signers = generate_signers_set(&env, 5, signers.domain_separator.clone());
+
+    assert_contract_err!(
+        client.try_rotate_signers(&new_signers.signers, &proof, &true),
+        ContractError::InvalidSignatures
+    );
+}
+
+#[test]
+fn fail_validate_proof_threshold_not_met() {
+    let (env, signers, client) = setup_env(randint(0, 10), randint(1, 10));
+
+    let mut total_weight = 0u128;
+
+    let (message, _) = generate_test_message(&env);
+    let messages = vec![&env, message.clone()];
+    let msg_hash = get_approve_hash(&env, messages.clone());
+
+    // let msg_hash: BytesN<32> = BytesN::random(&env);
+    let mut proof = generate_proof(&env, msg_hash.clone(), signers);
+
+    // Modify signatures to make them invalid
+    let mut new_signers = Vec::new(&env);
+    for ProofSigner { signer, signature } in proof.signers {
+        total_weight += signer.weight;
+
+        if total_weight < proof.threshold {
+            new_signers.push_back(ProofSigner { signer, signature });
+        } else {
+            new_signers.push_back(ProofSigner {
+                signer,
+                signature: ProofSignature::Unsigned,
+            });
+        }
+    }
+    proof.signers = new_signers;
+
+    // should panic, all signatures are valid but total weight is below threshold
+    assert_contract_err!(
+        client.try_approve_messages(&messages, &proof),
+        ContractError::InvalidSignatures
+    );
+}
+
+#[test]
+fn fail_validate_proof_invalid_signer_set() {
+    let (env, signers, client) = setup_env(randint(0, 10), randint(1, 10));
+
+    let new_signers = generate_signers_set(&env, randint(1, 10), signers.domain_separator.clone());
+
+    let msg_hash: BytesN<32> = BytesN::random(&env);
+    let mut proof = generate_proof(&env, msg_hash.clone(), signers);
+
+    let new_proof = generate_proof(&env, msg_hash.clone(), new_signers.clone());
+
+    proof.signers = new_proof.signers;
+
+    // validate_proof should panic, signatures do not match signers
+    assert_contract_err!(
+        client.try_rotate_signers(&new_signers.signers, &proof, &true),
+        ContractError::InvalidSignersHash
+    );
+}
+
+#[test]
+fn fail_validate_proof_threshold_overflow() {
+    let (env, mut signers, client) = setup_env(randint(0, 10), randint(1, 10));
+
+    let last_index = signers.signers.signers.len() - 1;
+
+    // get last signer and modify its weight to max u128 - 1
+    if let Some(mut last_signer) = signers.signers.signers.get(last_index) {
+        last_signer.weight = u128::MAX - 1;
+        signers.signers.signers.set(last_index, last_signer);
+    }
```

Comment:
```suggestion
    let ProofSigner { signer, weight } = signers.signers.signers.pop_back_unchecked() {
    signers.signers.signers.push_back(ProofSigner { signer, weight: u128::MAX });
```
use helper methods to make it cleaner. use u128::MAX itself to test overflow